Configurable Proxy settings

verdan opened the pull request:
The goal of the implementation is to make the proxy of amundsen metadata service configurable in a generic way. Need your thoughts about this. 

P.S: I haven't removed the neo4j settings yet, and only introduced the new changes along with the implementation of new proxy in the table details API. That's why the tests are also failing. 

cc: @feng-tao @jinhyukchang

verdan commented on metadata_service/proxy/base_proxy.py:
```diff
@@ -0,0 +1,6 @@
+class BaseProxy:
```
We can make this an interface, enforcing other clients to implement all functions needed, with same parameters.

verdan commented on metadata_service/config.py:
```diff
@@ -1,10 +1,23 @@
+
+
+PROXY_CLIENTS = {
+    'ATLAS': 'metadata_service.proxy.neo4j_proxy.Neo4jProxy',
```
This is temporary and will change once I'll implement the Atlas proxy.

jinhyukchang commented on setup.py:
```diff
@@ -2,6 +2,9 @@
 __version__ = '1.0.1'
 
+# Dependencies to make Amundsen work with Apache Atlas
+atlas = ['atlasclient>=0.1.2']
```
Just wonder how much of dependency `atlasclient` would bring. Brining both neo4j and atlas dependencies would be unnecessary but if it's light it should be fine.

feng-tao replied:
agree

feng-tao replied:
but I think @verdan brought that the atlasclient dependency is not very well maintained(https://amundsenworkspace.slack.com/archives/CGFBVT23V/p1552426276000400?thread_ts=1551375775.001000&cid=CGFBVT23V)?

verdan replied:
at the moment, yes we don't know what else would be needed to make `atlas` work with amundsen, therefore, I thought to keep it separate for now, and will combine neo4j and atlasclient, if this is the only dependency after the whole integration of atlas with amundsen. Thoughts? I can add a TODO here. 

@feng-tao true. The problem with atlasclient package is mainly the documentation, with incorrect code snippets/examples. So you need to hit and try or debug the functions in order to get the proper results. But this package also has some nice serializers to convert relations into JSON properly. Therefore I suggest to use this package and I will make sure to open PRs to update the documentation and refactor the library along the way.

feng-tao commented:
But the abstraction looks good!

verdan commented:
Nice, will update the APIs and test cases accordingly ASAP to avoid any conflicts. 👍

verdan commented:
So, after making the proxy client generic, the code gives the error that a positional argument `self` is missing. I can suppress that in PyCharm using `# noinspection PyArgumentList`, but wondering if there is an argument for mypy to disable that? 
```
mypy --ignore-missing-imports --follow-imports=skip --strict-optional --warn-no-return metadata_service
metadata_service/api/user.py:34: error: Missing positional argument "self" in call to "get_user_detail" of "BaseProxy"
metadata_service/api/tag.py:29: error: Too few arguments for "get_tags" of "BaseProxy"
metadata_service/api/table.py:94: error: Missing positional argument "self" in call to "get_table" of "BaseProxy"
metadata_service/api/table.py:112: error: Missing positional argument "self" in call to "add_owner" of "BaseProxy"
metadata_service/api/table.py:124: error: Missing positional argument "self" in call to "delete_owner" of "BaseProxy"
metadata_service/api/table.py:152: error: Missing positional argument "self" in call to "get_table_description" of "BaseProxy"
metadata_service/api/table.py:170: error: Missing positional argument "self" in call to "put_table_description" of "BaseProxy"
metadata_service/api/table.py:199: error: Missing positional argument "self" in call to "add_tag" of "BaseProxy"
metadata_service/api/table.py:220: error: Missing positional argument "self" in call to "delete_tag" of "BaseProxy"
metadata_service/api/system.py:19: error: Too few arguments for "get_latest_updated_ts" of "BaseProxy"
metadata_service/api/popular_tables.py:30: error: Too few arguments for "get_popular_tables" of "BaseProxy"
metadata_service/api/column.py:31: error: Missing positional argument "self" in call to "put_column_description" of "BaseProxy"
metadata_service/api/column.py:47: error: Missing positional argument "self" in call to "get_column_description" of "BaseProxy"
Makefile:16: recipe for target 'mypy' failed
make: *** [mypy] Error 1
```

feng-tao commented:
@verdan , Will take a look , also cc @jinhyukchang  and see if he knew about this in mypy. In the mean time, could you rebase the pr with master? Thanks.

jinhyukchang commented:
> @verdan , Will take a look , also cc @jinhyukchang and see if he knew about this in mypy. In the mean time, could you rebase the pr with master? Thanks.

@verdan Let me take a look as well and get back to you.

jinhyukchang commented on metadata_service/proxy/__init__.py:
```diff
@@ -0,0 +1,37 @@
+_proxy_client_lock = Lock()
+
+
+def get_proxy_client() -> Type[BaseProxy]:
```
@verdan Could you replace `Type[BaseProxy]` to `BaseProxy`? I found mypy error is gone after this

verdan replied:
nice catch!! I used the `Type[]` in the wrong way here. 👍

verdan commented:
@jinhyukchang ready for another pass.

feng-tao commented on metadata_service/config.py:
```diff
@@ -1,10 +1,22 @@
+PROXY_CLIENT = 'PROXY_CLIENT'
+
+
+PROXY_CLIENTS = {
```
do you want to add a todo to update this part once atlas client is implemented?

verdan replied:
I already have that task in my JIRA ;) 
The only thing now I'm considering is to remove the NEO4J variables now, as the generic config variables can now be used instead.

feng-tao commented:
thanks @verdan for the hard work 👍

jinhyukchang commented on metadata_service/api/user.py:
```diff
@@ -166,8 +165,8 @@ def put(self, user_id: str, resource_type: str, table_uri: str) -> Iterable[Unio
         try:
-            self.neo4j.delete_owner(table_uri=table_uri,
-                                    owner=user_id)
+            # noinspection PyArgumentList
```
Nit: remove commented code

jinhyukchang commented on setup.py:
```diff
@@ -2,6 +2,9 @@
 
 __version__ = '1.0.2'
```
@verdan Could you increase the version so that it can be released to Pypi?

feng-tao commented:
@verdan , let us know once you bump the version and remove the comment. And I think the pr is good to go :)

verdan commented:
Okay guys, one last commit, I just bumped the version, removed the comment and also removed the Neo4j specific config variables that are not needed anymore. @feng-tao @jinhyukchang

feng-tao commented:
LGTM @verdan

feng-tao commented:
merge the pr, thanks @verdan

verdan commented:
Should we manually squash the commits before merging or simply use the Github's squash merge feature ?

feng-tao commented:
yeah, I thought it default will squash the commit like Airflow....But let's remember to  squash all the commits into single commit before merge next time :)
